계산기 [STEP1] sehong

sehoong0429 opened the pull request:
안녕하세요 그루트!🌳 @Groot-94 
세홍이라고 합니다!
아직은 실력이 너무 부족해서 시간이 많이 소요되어 늦어진 점 죄송합니다.
이번 프로젝트 진행동안 잘 부탁드립니다. 🥕🥕

---
## 고민했던 점 및 궁금한 점 
<img width="500" alt="image" src=https://user-images.githubusercontent.com/67987230/214822721-a10dc6d5-2acd-4d85-89ce-186c94a50ef4.png>


1. UML을 코드를 작성하기 전에 그려야 할지 아니면 코드 구현을 마친 후에 그려야할지 고민하다가 머릿속에 그려지는 게 없어서 코드를 구현하고 작성하게 되었습니다. 오히려 좀 더 정리된 느낌을 받으면서 내가 어떤 타입을 만들어서 구현했는지 돌아보는 것도 좋았지만, 아직 UML 화살표 방향이나 점선 실선 등의 의의 파악을 하는 게 부족해서 공부를 더 해야 한다고 느꼈고, 사용의의 중 하나인 계획을 해서 진행과정을 만들어 나가야 하는 `설계`에 맞지 않는 부분이 있는 것 아닌가 하는 고민이 들었습니다. 

- UML은 Protocol, Struct, Class 각각의 색상을 달리하여 그렸습니다. 
- `CalculatatorItemQueue`는 `CalculatorItem`프로토콜을 채택하고 있기 때문에 점선(실체화)으로 표현했습니다. 
- `calculatorItemQueue`는 `CalculatorItemNode`타입을 활용하고 참조하고 있기 있기 때문에 실선(방향성이 있는 연관)으로 표현했습니다. 


2. CalculatorItemQueue를 구현할 때 `class`를 이용해야 할지 `struct`를 이용해야 할지에 대해서 고민했습니다.
class로 생성한 Node타입의 경우는 CalculatorItemQueue가 해당 타입을 참조하고 사용하고 있고 데이터의 잦은 변경이 있기때문에 class가 적합하다고 생각하였고, CaculatorItemQueue를 struct로 구현한 이유는 먼저 상속이 일어나지 않는다는 것을 들어볼 수 있습니다.. 그리고 struct를 내부 프로퍼티의 값을 변경할 일이 없고 단순하게 인스턴스 메소드만을 사용할 목적을 가질때도 좋다고 했는데 mutating을 사용하여 함수를 이용한 내부 프로퍼티의 값이 변동 되기도 합니다. 이를 위해서 class를 사용하는 게 문제가 될까요? 참조타입 값타입에 대한 부분을 저번 프로젝트에서 찾아보면서 고민을 해봤지만 queue는 class struct 어떤걸 사용해서 구현하는게 좋을지 궁금합니다.

3. 테스트 코드에서는 `private` 적용시 접근이 불가한데 테스트를 위해서는 `private` 적용 시키는 부분은 어떻게 해야할까요?
테스트를 위해서 private을 삭제시켰다가 다시 수정하는 방식은 너무 비효율적이라고 생각이 들어서 당장은 모두 접근이 가능한 방식으로 수정해두었으나 다른 방식이 있는지 궁금합니다.

sehoong0429 committed:
feat:
1. protocol CalculatorItem 파일 생성
2. CalculatorItemQueue 파일 생성
3. test 파일 생성

sehoong0429 committed:
feat: CalculatorItemQueue에서 사용할 Node타입 생성

sehoong0429 committed:
refactor:
1. 코드에 대한 간단한 설명 주석 추가
2. class, struct 첫번째 줄 마지막줄 띄우기
feat:
Queue 타입 구현하기
1. 해당 리스트의 첫번째(head), 마지막(last)값을 알려줄 변수 선언
2. 해당 리스트의 빈 배열 여부 판단 할 수 있는 isEmpty 구현
3. 해당 리스트 마지막 부분에 값을 추가 해주는 enqueue 함수구현
4. 해당 리스트의 첫번째 값을 삭제 해주는 dequeue 함수구현
5. 해당 리스트의 모든 값을 삭제할 수 있는 removeall 함수구현

sehoong0429 committed:
feat:
1. calculatorItemQueue파일에 대한 모든 test 케이스 구현
2. test케이스 구현할때 private접근 제어자로 인해 접근 하지 못해서 public으로 수정
3. calculatorItemQueue파일에 dequeue함수 테스트시 에러 발생으로 로직 수정

sehoong0429 committed:
chore: @testable import 삭제
삭제한 이유 File 'CalculatorTests.swift' is part of module 'Calculator'; ignoring import 라는 경고가 떴는데 이미 나는 target Membership으로 해당 테스트와 파일을 연결해줘서 필요가 없음.

sehoong0429 committed:
chore: UML을 정리하면서 node타입은 분리해줘도 좋을 것 같아서 분리함.

sehoong0429 committed:
docs: UML 추가

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+//  Created by Seoyeon Hong on 2023/01/26.
+//
+//MARK: Queue 타입 구현
+struct CalculatorItemQueue<T>: CalculateItem {
```
빈 프로토콜을 채택하고 있는데 그 이유는 무엇일까요?🤔

sehoong0429 replied:
해당 부분은 제가 프로젝트를 정확하게 숙지 후 진행하지 못했다는 것을 확인 후 수정하였습니다. 
프로젝트에서는 `CalculatorItemQueue`가 가진 `요소`가 `CalculateItem` 프로토콜을 채택해야합니다.

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+//MARK: Queue 타입 구현
+struct CalculatorItemQueue<T>: CalculateItem {
+    
+    var head: Node<T>?
```
현재 노드가 큐에 있는 상황인데 `노드는 큐를 구현하는데 필요한 List에 있어야 할 것 같습니다.` 이 부분은 프로젝트 내용을 다시 확인해보셔야 할 것 같습니다.

sehoong0429 replied:
제가 노드와 리스트에 대해서 정확하게 숙지하지 못한 뒤에 진행했습니다.
해당 부분 전체 삭제하였습니다.

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+        return node
+    }
+    
+    var isEmpty: Bool {
```
연산 프로퍼티와 함수를 사용하시는 기준이 궁금합니다.🤔

sehoong0429 replied:
제가 코드내에서 함수를 사용했던 이유는 
구현되는 함수의 로직이 복잡하거나 매 작업마다 다른 값을 반환할 때 쓰는 것이 유용하다고 생각해서입니다. 

연산 프로퍼티로 사용한 isEmpty의 경우에는 해당 queue에 값이 있는지 없는지 확인하는 단순한 작업을 실행하고, 매번 true, false의 두가지 경우의 단순한 값이 반환되기 때문에 사용하게 되었습니다.

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+    var isEmpty: Bool {
+        if head == nil {
+            return true
+        }else {
```
`} else {` 가 조금 더 자연스러운 컨벤션 같네요 ㅎㅎ

sehoong0429 replied:
감사합니다. 해당 부분 수정하였습니다!

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+    }
+    
+    mutating func enqueue(_ element: T) {
+        if let lastNode = last{
```
이 부분도 컨벤션 수정이 필요해보입니다!

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+    }
+    
+    mutating func dequeue() -> T? {
+        if head == nil { return nil }
```
if 대신 guard 를 사용하면 어떤 차이가 있을까요?

sehoong0429 replied:
문서에 따르면 if와 guard의 차이는 
guard가 항상 else문을 갖는 것입니다!

가독성 측면에서 더 좋아질 것 같습니다.

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+    mutating func dequeue() -> T? {
+        if head == nil { return nil }
+        
+        guard let nowhead = head else {
```
now 보단 current가 어떠신가요?

sehoong0429 replied:
네이밍에 대해서 조금 더 고민하고 사용헤보도록 하겠습니다.

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+}
+
+//MARK: Node 타입 생성
+class Node<T> {
```
Node는 상속이 될 가능성이 있을까요?

sehoong0429 replied:
Node는 상속 될 가능성이 없어서 
해당 부분은 `final`을 사용하여 상속을 막을 수 있을 것 같습니다!

Groot-94 commented on Calculator/Calculator/CalculatorItemQueue.swift:
```diff
@@ -0,0 +1,70 @@
+        guard head != nil else {
+            return
+        }
+        
```
이 부분의 공백은 의도하셨을까요?? 현재 작성하신 코드 컨벤션의 기준은 함수 내부에 공백이 없어보입니다.

Groot-94 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,122 @@
+        sut = nil
+    }
+    
+    func test_head접근시_첫번째노드를반환한다() {
```
given, when, then 주석으로 테스트 내부의 조건을 나눠주시면 가독성이 좋아질 것 같습니다.

sehoong0429 replied:
감사합니다! given, when, then을 사용하여 내부의 조건을 나누는 방식으로 수정했는데 질문이 있습니다!
테스트 코드의 이름은 영어로 작성하면 안되는 부분일까요? 
아래와 같이 간단하게 isEmpty, isNotEmpty 정도로 함수를 표현하는것에 문제가 있는지 아니면 다른 좋은 방식이 있는지 궁금합니다. 
```swift 
func test_IsEmpty() {
        //when
        sut.enqueue(1)
        sut.dequeue()
        
        //then
        XCTAssertTrue(sut.isEmpty)
    }
```
```swift
    func test_IsNotEmpty() {
        //when
        sut.enqueue(1)
        
        //then
        XCTAssertFalse(sut.isEmpty)
    }
```

Groot-94 replied:
영어로 표현해도 문제 없을것 같습니다. 하지만 isEmpty만 작성하기 보단, 어떤 케이스를 테스트 하는지 명시해주면 좋을 것 같네요.

sehoong0429 replied:
해당 명시하는 방식으로 수정 완료하였습니다! 감사합니다.

Groot-94 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,122 @@
+        sut.enqueue(2)
+        
+        let expectation = 1
+        let result = sut.head?.data
```
큐의 헤더에 직접적으로 접근해서 데이터를 확인하는 부분은 큐와 무관해보입니다.
1. 큐 자료구조에 헤더는 없어도 될 것 같습니다. (이 부분은 리스트 관련해서 코멘트 남겼습니다.)
2. 1번의 문제를 해결하셨다 가정하고, `sut.list.head?.data` 와 같이 접근했다면 큐의 리스트에 있는 헤드에 직접 접근해서 값을 확인하게 되는데 이 부분도 문제가 있다고 생각합니다. 
> 큐를 사용하는 사람은 큐가 데이터를 어떤 방식으로 저장하고 있는지는 관심이 없을 것 같습니다.  단순하게 큐의 맨 앞에 어떤 값이 들어있는지만 알 수 있도록 해주는게 어떨까요??🤔 
> 아직은 어려울 수 있으니 고민해보시고 이 부분이 이해가 가지 않으시면 디엠주세요 ㅎㅎ

Groot-94 commented on Calculator/CalculatorTests/CalculatorTests.swift:
```diff
@@ -0,0 +1,122 @@
+        XCTAssertEqual(expectation, result)
+    }
+    
+    func test_isEmpty접근할때_요소가_있으면_Bool을반환한다() {
```
무조건 결과는 Bool 타입이기 때문에 성공, 실패와 상관없는 케이스라고 느껴집니다.  Bool을 반환한다는 부분을 적절한 테스트 케이스로 수정해야 할 것 같습니다.

Groot-94 commented:
안녕하세요 :)

저도 잘 부탁드립니다. 현재 작성하신 커밋을 보면 TDD를 사용하셨는지 의문이 듭니다.🤔🤔
Unit Test와 TDD는 다른 의미라서 꼭 경험해보시면 좋을 것 같아요.

### 고민했던 점 및 궁금한 점
#### UML을 코드를 작성하기 전에 그려야 할지 아니면 코드 구현을 마친 후에 그려야할지 
- 저도 UML을 그려놓고 코드를 작성하는게 좋다고 생각합니다. 설계하는 능력을 키우기 위해선 UML을 그리고 코드를 작성하는 연습을 하시면 좋을 것 같습니다. 만약 UML만 작성하는게 어렵다면 단순하게 필요한 기능 리스트를 작성하시는 방법도 있을 것 같습니다.
#### CalculatorItemQueue를 구현할 때 class를 이용해야 할지 struct를 이용해야 할지에 대해서 고민했습니다.
- 현재는 [Choosing Between Structures and Classes](https://developer.apple.com/documentation/swift/choosing-between-structures-and-classes) 문서를 참고해서 선택하시면 될 것 같습니다. 현재는 참조와 값의 메모리에 대한 자세한 학습이 어려울 수 있습니다. 이 부분은 추후에 나중에 학습하실 내용입니다. ㅎㅎ
#### 테스트 코드에서는 private 적용시 접근이 불가한데 테스트를 위해서는 private 적용 시키는 부분은 어떻게 해야할까요?
- 단순하게 테스트를 위해서 코드를 수정하는 부분은 말씀하신 것 처럼 비효율적이라 생각합니다. 
- 코멘트에서도 남겼지만, private를 해줘야 하는 부분과 아닌 부분을 먼저 구분하는 기준을 고민해보시면 좋을 것 같습니다.
- 어떤 타입을 만드실 때 먼저 그 타입이 어떤 기능을 하는지, 그 기능을 하기 위해서 어떤 메서드를 가져야 하는지를 생각하시고, 외부에서 사용될 기능, 내부에서 사용될 기능을 구분한 상태에서 은닉화를 고민하면 좋을 것 같네요.

sehoong0429 committed:
feat:
1. CalculatorTest 파일 생성 및 extenstion int 추가
2. DoubleStack Queue타입으로 변경
3. Queue타입 내에 queue가 비어있는지 확인 할 수 있는 isEmpty 프로퍼티 생성 및 isEmpty & isNotEmpty 테스트 진행
4. Queue타입 내에 queue가 가지고 있는 첫번째 값을 출력하는 peek 프로퍼티 생성
5. Queue타입 내에 queue에 값을 추가할 수 있는 enqueue 함수 생성 및 enqueue함수 테스트 진행
6. Queue타입 내에 queue에 값을 삭제할 수 있는 dequeue 함수 생성 및 dequeue함수 테스트 진행

chore:
node파일 삭제(queue 구현방식 변경)
코드 컨벤션 수정
test파일 각각 given,when,then 주석추가

sehoong0429 committed:
docs: UML 수정

sehoong0429 committed:
refactor: 어떤 테스트를 진행한건지 테스트 명시해주기
